# Introduction

k-Nearest-Neighbors (kNN) is a simple technique for classification. The idea behind
it is that similar data points should have the same class, at least most of the time.
This method is very intuitive and has proven itself in many domains including
recommendation systems, anomaly detection, image/text classification and more.

The dataset we use contains information collected from unicorn ride data. The features are measurements like health points, magic points, temperature min, temperature max, and precipitation, while the label defines if too many health points were used for each ride. The question we have decided to answer is, "Given the distance and weather (temp min, temp max, precipitation), were too many health points consumed?"

We will use Amazon SageMaker's python SDK in order to train a kNN classifier in its simplest setting. We explain uploading data to Amazon S3, training a model, and using Lambda for serverless inference.


## Dataset

We're going to work with the unicorn ride dataset we uploaded to our raw S3 bucket. After you uploaded the file, a Lambda function added a few new features: temperature min, temperature max, and precipitation amount. The Lambda function also added a boolean (zero or one) label indicating if we think too many health points were used. Our current calculation is: distance > health_points * 100

In [93]:
%%time

import boto3
import pandas as pd
import io
import csv
import os

bucket = 'wildrydes-machine-learning-transformeddatabucket-1squv9nmot3md' # modify to your bucket name
prefix = 'processed'

s3 = boto3.resource('s3')
bucket = s3.Bucket(bucket)
objects = bucket.objects.filter(Prefix=prefix)

folder = '/home/ec2-user/SageMaker/'
for obj in objects:
    filename = os.path.join(folder,obj.key)
    os.makedirs(os.path.dirname(filename), exist_ok=True)
    bucket.download_file(obj.key, filename)

CPU times: user 1min 23s, sys: 5 s, total: 1min 28s
Wall time: 10min 4s


Let's get a count of how many files were downloaded...

In [95]:
%%bash
ls processed | wc -l

10417


Now that our files have downloaded to the notebook instance, let's load them into a dataframe.

In [115]:
import glob

filenames = glob.glob(folder + "/processed/*.csv")

list = []
for filename in filenames:
    list.append(pd.read_csv(filename))
frame = pd.concat(list, axis=0, ignore_index=True)
frame['statustime'] = pd.to_datetime(frame['statustime']).dt.to_period('D')
display(frame)

distance  healthpoints   latitude  longitude  magicpoints       name  \
0      293.858685          6789  43.259144 -67.758579         9362  Shadowfax   
1      216.607629          3363  42.149429 -70.541869          742  Shadowfax   
2      170.830697          5870  42.074028 -70.729197         1954  Shadowfax   
3      236.167542          6917  43.156521 -68.018071         2542  Shadowfax   
4      180.541236          8633  41.695710 -71.665760         9311  Shadowfax   
5      174.486841           513  47.279425 -57.223728         9974  Shadowfax   
6      191.769035          4934  41.493068 -72.165154         2571  Shadowfax   
7      188.086101          3773  44.246213 -65.239965         8714  Shadowfax   
8      136.987470          5321  41.283103 -72.680951         4020  Shadowfax   
9       92.887627          3282  43.850283 -66.255221         5516  Shadowfax   
10     285.575483          2310  44.307771 -65.081502         7872  Shadowfax   
11     250.121132          2140  41.139758 -73.032136         4858  Shadowfax   
12     173.838338          8833  45.658494 -61.561695         8164  Shadowfax   
13      53.791859          5903  44.669422 -64.147173         9873  Shadowfax   
14     207.375663          7048  47.418436 -56.845584         6650  Shadowfax   
15      36.274161           454  41.545984 -72.034898         6698  Shadowfax   
16     253.726557          8198  42.068983 -70.741724         9577  Shadowfax   
17     225.057089          8987  42.876434 -68.724094         7988  Shadowfax   
18     275.841135          1367  43.628451 -66.821105         7433  Shadowfax   
19      55.721991           382  43.009930 -68.387989         3608  Shadowfax   
20     133.006617          9096  43.926621 -66.059998         6603  Shadowfax   
21     166.867811          3994  42.502818 -69.660910         5147  Shadowfax   
22      33.696242          1308  41.190896 -72.906940         5964  Shadowfax   
23     258.458768          2140  42.684347 -69.206443         1737  Shadowfax   
24      49.137636          1881  41.394523 -72.407446         3545  Shadowfax   
25     171.820291          9880  43.914025 -66.092227         8399  Shadowfax   
26     222.838085          2634  44.570063 -64.404446         3189  Shadowfax   
27     130.486763          1978  43.213561 -67.873893         5549  Shadowfax   
28      53.639999          7993  42.002796 -70.905962          268  Shadowfax   
29      98.014301          3547  41.568494 -71.979457         1950  Shadowfax   
...           ...           ...        ...        ...          ...        ...   
10387  216.562951          4826  40.988400 -73.402120         2529  Shadowfax   
10388   30.101617          8923  43.832092 -66.301703         3675  Shadowfax   
10389  114.765575          6774  47.372839 -56.969731         7057  Shadowfax   
10390   88.852974          8327  41.209193 -72.862121         7134  Shadowfax   
10391  135.788475          3852  42.774673 -68.979812         8269  Shadowfax   
10392  245.671940          6265  45.952907 -60.783295         1409  Shadowfax   
10393   35.798630          9781  44.510967 -64.557258         6712  Shadowfax   
10394  170.832217          3279  43.022377 -68.356614         3517  Shadowfax   
10395  153.317408          7147  45.538380 -61.878090         7579  Shadowfax   
10396   72.368560          2791  43.561151 -66.992370         1279  Shadowfax   
10397  205.924125          9499  44.943115 -63.436201         1323  Shadowfax   
10398   84.102445          3528  42.161220 -70.512554         3989  Shadowfax   
10399  277.129214          2331  47.691101 -56.100965         6814  Shadowfax   
10400  263.586773            23  41.385219 -72.430303         9289  Shadowfax   
10401  271.283886          9991  46.692929 -58.808295         2535  Shadowfax   
10402  268.825838          3397  48.115137 -54.935185         8121  Shadowfax   
10403  207.746091          3448  42.608274 -69.397058         2232  Shadowfax   
10404  225.762687          9098  43.4044

Let's load our ground station weather data into a dataframe.

In [121]:
import boto3
import pandas as pd
import io
import csv

bucket = 'wildrydes-machine-learning-transformeddatabucket-1squv9nmot3md' # modify to your bucket name
prefix = 'nygroundstationdata'

s3 = boto3.resource('s3')
bucket = s3.Bucket(bucket)
objects = bucket.objects.filter(Prefix=prefix)
for obj in objects:
    if obj.key.endswith('csv'):
        ny = pd.read_csv(obj.get()['Body'], index_col="id")
ny['year_date'] = pd.to_datetime(ny['year_date'], format='%Y%m%d').dt.to_period('D')
display(ny)

year_date element  data_value m_flag  q_flag s_flag  obs_time
id                                                                         
USW00094728  1869-01-01    TMAX         -17    NaN     NaN      Z       NaN
USW00094728  1869-01-01    TMIN         -72    NaN     NaN      Z       NaN
USW00094728  1869-01-01    PRCP         191    NaN     NaN      Z       NaN
USW00094728  1869-01-01    SNOW         229    NaN     NaN      Z       NaN
USW00094728  1869-01-02    TMAX         -28    NaN     NaN      Z       NaN
USW00094728  1869-01-02    TMIN         -61    NaN     NaN      Z       NaN
USW00094728  1869-01-02    PRCP           8    NaN     NaN      Z       NaN
USW00094728  1869-01-02    SNOW           0    NaN     NaN      Z       NaN
USW00094728  1869-01-03    TMAX          17    NaN     NaN      Z       NaN
USW00094728  1869-01-03    TMIN         -28    NaN     NaN      Z       NaN
USW00094728  1869-01-03    PRCP           0      T     NaN      Z       NaN
USW00094728  1869-01-03    SNOW           0    NaN     NaN      Z       NaN
USW00094728  1869-01-04    TMAX          28    NaN     NaN      Z       NaN
USW00094728  1869-01-04    TMIN          11    NaN     NaN      Z       NaN
USW00094728  1869-01-04    PRCP          46    NaN     NaN      Z       NaN
USW00094728  1869-01-04    SNOW           0    NaN     NaN      Z       NaN
USW00094728  1869-01-05    TMAX          61    NaN     NaN      Z       NaN
USW00094728  1869-01-05    TMIN          28    NaN     NaN      Z       NaN
USW00094728  1869-01-05    PRCP          13    NaN     NaN      Z       NaN
USW00094728  1869-01-05    SNOW           0    NaN     NaN      Z       NaN
USW00094728  1869-01-06    TMAX          33    NaN     NaN      Z       NaN
USW00094728  1869-01-06    TMIN          11    NaN     NaN      Z       NaN
USW00094728  1869-01-06    PRCP           0    NaN     NaN      Z       NaN
USW00094728  1869-01-06    SNOW           0    NaN     NaN      Z       NaN
USW00094728  1869-01-07    TMAX          89    NaN     NaN      Z       NaN
USW00094728  1869-01-07    TMIN          17    NaN     NaN      Z       NaN
USW00094728  1869-01-07    PRCP           0    NaN     NaN      Z       NaN
USW00094728  1869-01-07    SNOW           0    NaN     NaN      Z       NaN
USW00094728  1869-01-08    TMAX         122    NaN     NaN      Z       NaN
USW00094728  1869-01-08    TMIN          44    NaN     NaN      Z       NaN
...                 ...     ...         ...    ...     ...    ...       ...
USW00094789  2012-01-19    TMAX          33    NaN     NaN      X    2400.0
USW00094789  2012-01-19    TMIN         -78    NaN     NaN      X    2400.0
USW00094789  2012-01-19    PRCP           0    NaN     NaN      X    2400.0
USW00094789  2012-01-19    SNOW           0    NaN     NaN      X       NaN
USW00094789  2012-01-19    SNWD           0    NaN     NaN      X       NaN
USW00094789  2012-01-19    AWND          47    NaN     NaN      W       NaN
USW00094789  2012-01-19    WDF2         190    NaN     NaN      X       NaN
USW00094789  2012-01-19    WDF5         180    NaN     NaN      X       NaN
USW00094789  2012-01-19    WSF2         103    NaN     NaN      X       NaN
USW00094789  2012-01-19    WSF5         134    NaN     NaN      X       NaN
USW00094728  2012-01-19    TMAX          28    NaN     NaN      X    2400.0
USW00094728  2012-01-19    TMIN         -67    NaN     NaN      X    2400.0
USW00094728  2012-01-19    PRCP           0    NaN     NaN      X    2400.0
USW00094728  2012-01-19    SNOW           0    NaN     NaN      X       NaN
USW00094728  2012-01-19    SNWD           0    NaN     NaN      X       NaN
USW00094728  2012-01-19    AWND          30    NaN     NaN      W       NaN
USW00094728  2012-01-19    WDF2         250    NaN     NaN      X       NaN
USW00094728  2012-01-19    WDF5         220    NaN     NaN      X       NaN
USW00094728  2012-01-19    WSF2          72    NaN     NaN      X       NaN
USW00094728  2012-01-19    WSF5         107    NaN     NaN    

As you can see we have weather data across the state of New York going back to 1869. We can focus on a certain date and data point...

In [79]:
ny_june_first_temp_min = ny.query('year_date == 20190601 and element == "TMIN"')
display(ny_june_first_temp_min)

year_date element  data_value m_flag  q_flag s_flag  obs_time
id                                                                        
USW00094789   20190601    TMIN         161    NaN     NaN      D    2400.0
USW00094728   20190601    TMIN         172    NaN     NaN      D    2400.0
USW00014732   20190601    TMIN         172    NaN     NaN      D    2400.0

Then grab that value from our result...

In [80]:
specific_station = ny_june_first_temp_min.loc["USW00094728"]
display(specific_station['data_value'])

172

In [101]:
test = ny.loc["USW00094728"].query('year_date == 20190601')
test.pivot(index='year_date', columns='element', values='data_value')

element    AWND  PRCP  SNOW  SNWD  TMAX  TMIN  WDF2  WDF5  WSF2  WSF5  WT08
year_date                                                                  
20190601     12     0     0     0   267   172   150   150    40    72     1

In [131]:
ny_pivot = ny.pivot_table(index=['year_date','id'], columns='element', values='data_value')
display(ny_pivot[['TMIN','TMAX','PRCP']])

element                  TMIN   TMAX   PRCP
year_date  id                              
1869-01-01 USW00094728  -72.0  -17.0  191.0
1869-01-02 USW00094728  -61.0  -28.0    8.0
1869-01-03 USW00094728  -28.0   17.0    0.0
1869-01-04 USW00094728   11.0   28.0   46.0
1869-01-05 USW00094728   28.0   61.0   13.0
1869-01-06 USW00094728   11.0   33.0    0.0
1869-01-07 USW00094728   17.0   89.0    0.0
1869-01-08 USW00094728   44.0  122.0    0.0
1869-01-09 USW00094728   33.0   89.0    0.0
1869-01-10 USW00094728    6.0   67.0    3.0
1869-01-11 USW00094728  -11.0    6.0    0.0
1869-01-12 USW00094728  -17.0   28.0  216.0
1869-01-13 USW00094728  -22.0   33.0    0.0
1869-01-14 USW00094728    0.0   56.0    0.0
1869-01-15 USW00094728   39.0   56.0   10.0
1869-01-16 USW00094728    0.0   33.0    0.0
1869-01-17 USW00094728  -17.0   17.0    0.0
1869-01-18 USW00094728  -33.0  -17.0    0.0
1869-01-19 USW00094728  -28.0   11.0   38.0
1869-01-20 USW00094728  -11.0   28.0    0.0
1869-01-21 USW00094728  -17.0   56.0    0.0
1869-01-22 USW00094728  -83.0  -17.0    0.0
1869-01-23 USW00094728  -67.0   39.0    0.0
1869-01-24 USW00094728   17.0   89.0    0.0
1869-01-25 USW00094728  -56.0   33.0    0.0
1869-01-26 USW00094728  -78.0   -6.0    0.0
1869-01-27 USW00094728  -50.0   44.0    0.0
1869-01-28 USW00094728   17.0   83.0    0.0
1869-01-29 USW00094728  -11.0   89.0    0.0
1869-01-30 USW00094728   50.0  122.0  119.0
...                       ...    ...    ...
2019-06-21 USW00014732  183.0  256.0  137.0
           USW00094728  183.0  250.0  135.0
           USW00094789  189.0  267.0  147.0
2019-06-22 USW00014732  172.0  267.0    0.0
           USW00094728  161.0  261.0    0.0
           USW00094789  172.0  278.0    0.0
2019-06-23 USW00014732  194.0  300.0    0.0
           USW00094728  178.0  283.0    0.0
           USW00094789  178.0  294.0    0.0
2019-06-24 US1NYNY0074    NaN    NaN    NaN
           USW00014732  211.0  300.0    0.0
           USW00094728  194.0  289.0    0.0
           USW00094789  172.0  283.0    0.0
2019-06-25 US1NYNY0074    NaN    NaN  107.0
           USW00014732  217.0  306.0  107.0
           USW00094728  222.0  289.0  218.0
           USW00094789  206.0  272.0   53.0
2019-06-26 US1NYNY0074    NaN    NaN   71.0
           USW00014732  206.0  322.0    0.0
           USW00094728  222.0  306.0    0.0
           USW00094789  206.0  300.0    0.0
2019-06-27 USW00014732  222.0  322.0    0.0
           USW00094728  217.0  317.0    0.0
           USW00094789  206.0  328.0    0.0
2019-06-28 USW00014732  217.0  339.0    0.0
           USW00094728  228.0  317.0    0.0
           USW00094789  206.0  311.0    0.0
2019-06-29 USW00014732  244.0  344.0   10.0
           USW00094728  222.0  328.0   20.0
           USW00094789  217.0  317.0    8.0

[135626 rows x 3 columns]

In [133]:
merge_df = frame.merge(ny_pivot, left_on=['statustime','groundstation'], right_on=['year_date','id'])
display(merge_df[['distance','groundstation','TMIN','TMAX','PRCP']])

distance groundstation   TMIN   TMAX   PRCP
0    97.822953   USW00014732  217.0  306.0  107.0
1    34.810117   USW00014732  217.0  306.0  107.0
2   276.712498   USW00014732  217.0  306.0  107.0
3   257.830391   USW00014732  217.0  306.0  107.0
4   286.616123   USW00014732  217.0  306.0  107.0
5   289.073448   USW00014732  217.0  306.0  107.0
6    33.417093   USW00014732  217.0  306.0  107.0
7   167.841328   USW00014732  217.0  306.0  107.0
8    40.487662   USW00014732  217.0  306.0  107.0
9   211.718501   USW00014732  217.0  306.0  107.0
10  231.232564   USW00014732  217.0  306.0  107.0
11  291.494703   USW00014732  217.0  306.0  107.0
12  254.374779   USW00014732  217.0  306.0  107.0
13  136.167010   USW00014732  217.0  306.0  107.0
14   92.873556   USW00014732  217.0  306.0  107.0
15   36.126334   USW00014732  217.0  306.0  107.0
16   52.071215   USW00014732  217.0  306.0  107.0
17   91.613965   USW00014732  217.0  306.0  107.0
18  183.396408   USW00014732  217.0  306.0  107.0
19  204.392494   USW00014732  217.0  306.0  107.0
20  291.109101   USW00014732  217.0  306.0  107.0
21  263.490131   USW00014732  217.0  306.0  107.0
22   87.571298   USW00014732  217.0  306.0  107.0
23  135.516297   USW00014732  217.0  306.0  107.0
24  199.141963   USW00014732  217.0  306.0  107.0
25  132.281413   USW00014732  217.0  306.0  107.0
26   38.320026   USW00014732  217.0  306.0  107.0
27  231.404493   USW00014732  217.0  306.0  107.0
28  110.333597   USW00014732  217.0  306.0  107.0
29   99.652938   USW00014732  217.0  306.0  107.0
..         ...           ...    ...    ...    ...
63  182.389022   USW00014732  217.0  306.0  107.0
64  257.126529   US1NYNY0074    NaN    NaN  107.0
65   83.422596   US1NYNY0074    NaN    NaN  107.0
66  110.947878   US1NYNY0074    NaN    NaN  107.0
67  148.397735   US1NYNY0074    NaN    NaN  107.0
68  287.544220   US1NYNY0074    NaN    NaN  107.0
69  156.422635   US1NYNY0074    NaN    NaN  107.0
70  267.573194   US1NYNY0074    NaN    NaN  107.0
71   72.291943   US1NYNY0074    NaN    NaN  107.0
72  141.865797   US1NYNY0074    NaN    NaN  107.0
73  146.468268   US1NYNY0074    NaN    NaN  107.0
74  281.412031   US1NYNY0074    NaN    NaN  107.0
75  268.212835   US1NYNY0074    NaN    NaN  107.0
76  181.417812   USW00094728  222.0  289.0  218.0
77  191.053637   USW00094728  222.0  289.0  218.0
78  148.779584   USW00094728  222.0  289.0  218.0
79  117.847265   USW00094728  222.0  289.0  218.0
80  228.085367   USW00094728  222.0  289.0  218.0
81  294.480575   USW00094728  222.0  289.0  218.0
82  118.936388   USW00094728  222.0  289.0  218.0
83  158.551836   USW00094728  222.0  289.0  218.0
84  196.525869   USW00094728  222.0  289.0  218.0
85   68.070355   USW00094728  222.0  289.0  218.0
86  258.575516   USW00094728  222.0  289.0  218.0
87  110.599779   USW00094728  222.0  289.0  218.0
88   92.025829   USW00094728  222.0  289.0  218.0
89   88.599055   USW00094728  222.0  289.0  218.0
90   32.557613   USW00094728  222.0  289.0  218.0
91  132.478372   USW00094728  222.0  289.0  218.0
92  188.689282   USW00094728  222.0  289.0  218.0

[93 rows x 5 columns]

## Pre-Processing the Data
Now that we have the raw data, let's process it. 
We'll first load the data into numpy arrays, and randomly split it into train and test with a 90/10 split.

In [ ]:
%%time

import numpy as np
import os

data_dir = "/tmp/"
processed_subdir = "standardized"
raw_data_file = os.path.join(data_dir, "raw", "transformed_data.csv")
train_features_file = os.path.join(data_dir, processed_subdir, "train/csv/features.csv")
train_labels_file = os.path.join(data_dir, processed_subdir, "train/csv/labels.csv")
test_features_file = os.path.join(data_dir, processed_subdir, "test/csv/features.csv")
test_labels_file = os.path.join(data_dir, processed_subdir, "test/csv/labels.csv")

# read raw data
print("Reading raw data from {}".format(raw_data_file))
raw = np.loadtxt(raw_data_file, delimiter=',', usecols=(0,1,4,10)) #,7,8,9,10))

# split into train/test with a 90/10 split
np.random.seed(0)
np.random.shuffle(raw)
train_size = int(0.9 * raw.shape[0])
train_features = raw[:train_size, :-1]
train_labels = raw[:train_size, -1]
test_features = raw[train_size:, :-1]
test_labels = raw[train_size:, -1]

## Upload to Amazon S3
Now, since typically the dataset will be large and located in Amazon S3, let's write the data to Amazon S3 in recordio-protobuf format. We first create an io buffer wrapping the data, next we upload it to Amazon S3. Notice that the choice of bucket and prefix should change for different users and different datasets

In [ ]:
import io
import sagemaker.amazon.common as smac

print('train_features shape = ', train_features.shape)
print('train_labels shape = ', train_labels.shape)

buf = io.BytesIO()
smac.write_numpy_to_dense_tensor(buf, train_features, train_labels)
buf.seek(0)

In [ ]:
import boto3
import os
import sagemaker

bucket = '' # modify to your bucket name
prefix = 'knn-blog-2018-04-17'
key = 'recordio-pb-data'

boto3.resource('s3').Bucket(bucket).Object(os.path.join(prefix, 'train', key)).upload_fileobj(buf)
s3_train_data = 's3://{}/{}/train/{}'.format(bucket, prefix, key)
print('uploaded training data location: {}'.format(s3_train_data))


It is also possible to provide test data. This way we can get an evaluation of the performance of the model from the training logs. In order to use this capability let's upload the test data to Amazon S3 as well

In [ ]:
print('test_features shape = ', test_features.shape)
print('test_labels shape = ', test_labels.shape)

buf = io.BytesIO()
smac.write_numpy_to_dense_tensor(buf, test_features, test_labels)
buf.seek(0)

boto3.resource('s3').Bucket(bucket).Object(os.path.join(prefix, 'test', key)).upload_fileobj(buf)
s3_test_data = 's3://{}/{}/test/{}'.format(bucket, prefix, key)
print('uploaded test data location: {}'.format(s3_test_data))


## Training

We take a moment to explain at a high level, how Machine Learning training and prediction works in Amazon SageMaker. First, we need to train a model. This is a process that given a labeled dataset and hyper-parameters guiding the training process,  outputs a model. Once the training is done, we set up what is called an **endpoint**. An endpoint is a web service that given a request containing an unlabeled data point, or mini-batch of data points, returns a prediction(s).

In Amazon SageMaker the training is done via an object called an **estimator**. When setting up the estimator we specify the location (in Amazon S3) of the training data, the path (again in Amazon S3) to the output directory where the model will be serialized, generic hyper-parameters such as the machine type to use during the training process, and kNN-specific hyper-parameters such as the index type, etc. Once the estimator is initialized, we can call its **fit** method in order to do the actual training.

Now that we are ready for training, we start with a convenience function that starts a training job.

In [ ]:
import matplotlib.pyplot as plt

import sagemaker
from sagemaker import get_execution_role
from sagemaker.predictor import csv_serializer, json_deserializer
from sagemaker.amazon.amazon_estimator import get_image_uri


def trained_estimator_from_hyperparams(s3_train_data, hyperparams, output_path, s3_test_data=None):
    """
    Create an Estimator from the given hyperparams, fit to training data, 
    and return a deployed predictor
    
    """
    # set up the estimator
    knn = sagemaker.estimator.Estimator(get_image_uri(boto3.Session().region_name, "knn"),
        get_execution_role(),
        train_instance_count=1,
        train_instance_type='ml.m5.2xlarge',
        output_path=output_path,
        sagemaker_session=sagemaker.Session())
    knn.set_hyperparameters(**hyperparams)
    
    # train a model. fit_input contains the locations of the train and test data
    fit_input = {'train': s3_train_data}
    if s3_test_data is not None:
        fit_input['test'] = s3_test_data
    knn.fit(fit_input)
    return knn

Now, we run the actual training job. For now, we stick to default parameters.

In [ ]:
hyperparams = {
    'feature_dim': 3,
    'k': 10,
    'sample_size': 8000,
    'predictor_type': 'classifier' 
}
output_path = 's3://' + bucket + '/' + prefix + '/default_example/output'
knn_estimator = trained_estimator_from_hyperparams(s3_train_data, hyperparams, output_path, 
                                                   s3_test_data=s3_test_data)

Notice that we mentioned a test set in the training job. When a test set is provided the training job doesn't just produce a model but also applies it to the test set and reports the accuracy. In the logs you can view the accuracy of the model on the test set.

## Setting up the endpoint

Now that we have a trained model, we are ready to run inference. The **knn_estimator** object above contains all the information we need for hosting the model. Below we provide a convenience function that given an estimator, sets up and endpoint that hosts the model. Other than the estimator object, we provide it with a name (string) for the estimator, and an **instance_type**. The **instance_type** is the machine type that will host the model. It is not restricted in any way by the parameter settings of the training job.

In [ ]:
def predictor_from_estimator(knn_estimator, estimator_name, instance_type, endpoint_name=None): 
    knn_predictor = knn_estimator.deploy(initial_instance_count=1, instance_type=instance_type,
                                        endpoint_name=endpoint_name)
    knn_predictor.content_type = 'text/csv'
    knn_predictor.serializer = csv_serializer
    knn_predictor.deserializer = json_deserializer
    return knn_predictor

In [ ]:
import time

instance_type = 'ml.m4.xlarge'
model_name = 'knn_%s'% instance_type
endpoint_name = 'knn-ml-m4-xlarge-%s'% (str(time.time()).replace('.','-'))
print('setting up the endpoint..')
predictor = predictor_from_estimator(knn_estimator, model_name, instance_type, endpoint_name=endpoint_name)

## Inference

Now that we have our predictor, let's use it on our test dataset. The following code runs on the test dataset, computes the accuracy and the average latency. It splits up the data into 100 batches, each of size roughly 500. Then, each batch is given to the inference service to obtain predictions. Once we have all predictions, we compute their accuracy given the true labels of the test set.


In [ ]:

batches = np.array_split(test_features, 100)
print('data split into 100 batches, of size %d.' % batches[0].shape[0])

# obtain an np array with the predictions for the entire test set
start_time = time.time()
predictions = []
for batch in batches:
    result = predictor.predict(batch)
    cur_predictions = np.array([result['predictions'][i]['predicted_label'] for i in range(len(result['predictions']))])
    predictions.append(cur_predictions)
predictions = np.concatenate(predictions)
run_time = time.time() - start_time

test_size = test_labels.shape[0]
num_correct = sum(predictions == test_labels)
accuracy = num_correct / float(test_size)
print('time required for predicting %d data point: %.2f seconds' % (test_size, run_time))
print('accuracy of model: %.1f%%' % (accuracy * 100) )

## Deleting the endpoint

We're now done with the example except a final clean-up act. By setting up the endpoint we started a machine in the cloud and as long as it's not deleted the machine is still up and we are paying for it. Once the endpoint is no longer necessary, we delete it. The following code does exactly that.

In [ ]:
def delete_endpoint(predictor):
    try:
        boto3.client('sagemaker').delete_endpoint(EndpointName=predictor.endpoint)
        print('Deleted {}'.format(predictor.endpoint))
    except:
        print('Already deleted: {}'.format(predictor.endpoint))

delete_endpoint(predictor)
            

## Conclusion

We've seen how to both train and host an inference endpoint for kNN. With absolutely zero tuning we obtain an accuracy of 92.2% on the covtype dataset. As a point of reference for grasping the prediction power of the kNN model, a linear model will achieve roughly 72.8% accuracy. There are several advanced issues that we did not discuss. In the next section we will deep-dive into issues such as run-time / latency, and tuning the model while taking into account both the accuracy and its run-time efficiency.

# Part 2: Deep dive, Tuning kNN 

Now that we managed to run a simple example we move to a more advanced one. We keep the objective of classifying the covtype data set but now see how to explore the different options of the kNN algorithm. In what follows we will review different solvers for the kNN problem, different instance types for hosting the model, and a verbose prediction option that allows tuning the k value of the model after training

## Train and Predict in Amazon SageMaker

Below we copy, for convenience the function we set for training an estimator. The function has an optional value of the location of a test set. If one is given, the training job will apply the model on a test set and report the accuracy in the logs. This way, it is possible to obtain the quality of a model without setting up an endpoint and applying it to a test set. Though the function accepts hyper-parameters for the training job, we mention two hyper-parameters that we fixed inside the function.
* **train_instance_count**: This is the number of machines used for the training process. For large datasets we may want to use multiple machines. In our case the entire dataset is rather small (~100MB) so we use only a single machine.
* **train_instance_type**: This is the machine type used for training. We use the recommended type for training of 'ml.m5.2xlarge'. In cases where we have a large test set, it may be preferable to use a GPU machine or a CPU compute optimized machine such 'ml.c5.2xlarge' as the compute resource required for running inference on the test set may be considerable.


In [ ]:
import matplotlib.pyplot as plt

import sagemaker
from sagemaker import get_execution_role
from sagemaker.predictor import csv_serializer, json_deserializer
from sagemaker.amazon.amazon_estimator import get_image_uri


def trained_estimator_from_hyperparams(s3_train_data, hyperparams, output_path, s3_test_data=None):
    """
    Create an Estimator from the given hyperparams, fit to training data, 
    and return a deployed predictor
    
    """

    # set up the estimator
    knn = sagemaker.estimator.Estimator(get_image_uri(boto3.Session().region_name, "knn"),
        get_execution_role(),
        train_instance_count=1,
        train_instance_type='ml.m5.2xlarge',
        output_path=output_path,
        sagemaker_session=sagemaker.Session())
    knn.set_hyperparameters(**hyperparams)
    
    # train a model. fit_input contains the locations of the train and test data
    fit_input = {'train': s3_train_data}
    if s3_test_data is not None:
        fit_input['test'] = s3_test_data
    knn.fit(fit_input)
    return knn

The function above returns the estimator object after running the **fit** method. The object contains all the information needed to set up an endpoint. The function below gets as input a fitted estimator and uses it to set up an endpoint. An important parameter this function receives is **instance_type**. The endpoint can be set on any machine type, regardless of the one used in training. 

In kNN, inference can be somewhat slow as it involves computing the distance from the given feature vector to all data points. For that reason, one might want to set up a compute optimized endpoint either on CPU (c5.xlarge) or GPU (p2.xlarge). In what follows we will set endpoints for both machine types and compare their performances.

In [ ]:
def predictor_from_hyperparams(knn_estimator, estimator_name, instance_type, endpoint_name=None): 
    knn_predictor = knn_estimator.deploy(initial_instance_count=1, instance_type=instance_type,
                                        endpoint_name=endpoint_name)
    knn_predictor.content_type = 'text/csv'
    knn_predictor.serializer = csv_serializer
    knn_predictor.deserializer = json_deserializer
    return knn_predictor

## Launching the Training Jobs

We're now ready to launch the training jobs. We'll train two different models, based on two different [FAISS indexes](TODO: point to our hyperparameter documentation). One will have a **IndexFlatL2** index, meaning it will find the nearest neighbors with a brute force approach. The other will use an **IndexIVFPQ** index that both uses a cell-probe method to avoid computing all the distances, and a product quantization method that improves the run-time of each distance calculation. For both classifiers, we use the L2 distance. It's also possible to use cosine distance but we do not explore that option here. Intuitively, the first index is accurate but may suffer from larger latencies / slower throughput. The second index is faster but may give less accurate responses, resulting in a lower accuracy/F1 score. 

We stop to mention an important parameter for Amazon SageMaker kNN, the **sample_size**. This parameter determines how many points from the data set should be used for building the model. Using all the data points is tempting and definitely can't hurt the quality of the outputs but is often either infeasible or simply too costly. Moreover, it can often be unnecessary in the sense that you may get the same accuracy with 200K points as you would with 2M points. In that case, there is simply no need to build a model with 2M points. In this example we use a sample of 200K points out of the potential ~450K in the training set.

In [ ]:
hyperparams_flat_l2 = {
    'feature_dim': 54,
    'k': 100,
    'sample_size': 200000,
    'predictor_type': 'classifier' 
    # NOTE: The default distance is L2 and index is Flat, so we don't list them here
}
output_path_flat_l2 = 's3://' + bucket + '/' + prefix + '/flat_l2/output'
knn_estimator_flat_l2 = trained_estimator_from_hyperparams(s3_train_data, hyperparams_flat_l2, output_path_flat_l2, 
                                                           s3_test_data=s3_test_data)

In [ ]:
hyperparams_flat_l2_large = {
    'feature_dim': 54,
    'k': 100,
    'sample_size': 500000,
    'predictor_type': 'classifier' 
}
output_path_flat_l2_large = 's3://' + bucket + '/' + prefix + '/flat_l2_large/output'
knn_estimator_flat_l2_large = trained_estimator_from_hyperparams(s3_train_data, hyperparams_flat_l2_large, 
                                                                 output_path_flat_l2_large, s3_test_data=s3_test_data)

In [ ]:
hyperparams_ivfpq_l2 = {
    'feature_dim': 54,
    'k': 100,
    'sample_size': 200000,
    'predictor_type': 'classifier',
    'index_type': 'faiss.IVFPQ',
    'index_metric': 'L2',
    'faiss_index_ivf_nlists': 500, 
    'faiss_index_pq_m': '8'
    # number of distinct quantizers. In our case dim=54, so every quantizer will code up to 7 numbers.
    # Each block of (up to) 7 numbers will be coded into a single byte.
}
output_path_ivfpq_l2 = 's3://' + bucket + '/' + prefix + '/ivfpq_l2/output'
knn_estimator_ivfpq_l2 = trained_estimator_from_hyperparams(s3_train_data, hyperparams_ivfpq_l2, output_path_ivfpq_l2, 
                                                           s3_test_data=s3_test_data)

In [ ]:
hyperparams_ivfpq_l2_large = {
    'feature_dim': 54,
    'k': 100,
    'sample_size': 500000,
    'predictor_type': 'classifier',
    'index_type': 'faiss.IVFPQ',
    'index_metric': 'L2',
    'faiss_index_ivf_nlists': 700, 
    'faiss_index_pq_m': '8'
}
output_path_ivfpq_l2_large = 's3://' + bucket + '/' + prefix + '/ivfpq_l2_large/output'
knn_estimator_ivfpq_l2_large = trained_estimator_from_hyperparams(s3_train_data, hyperparams_ivfpq_l2_large, 
                                                                  output_path_ivfpq_l2_large, s3_test_data=s3_test_data)

## Evaluation

Before we set up the endpoints we take a moment to discuss ways of using them. The typical way of using an endpoint is the following: We feed it an input point and recieve a prediction. For kNN we additionally support another verbose API allowing for a more detailed response. With that API, instead of recieving only a prediction we recieve an ordered list of the labels corresponding to the nearest neighbors, along with their distances. 

A manual inspection of this list allows us to perform multiple tasks:
* Tune k: By taking a majority vote over a prefix of the list we obtain the prediction for multiple values of k
* Ranking / multi-label: Sometimes, we do not wish to predict a single class rather output a ranking of possible classes. An example for this would be image tagging where the labels are tags and the input vectors represent images.

In what follows we focus on tuning k with this API. We begin with an **evaluate** function that measures both the output quality and latency of an endpoint. The code below contains this function along with auxiliary functions used by it. For reporting the quality we use the verbose API to obtain the predictions for multiple k values. For the latency we use the non-verbose API.

In [ ]:
from scipy import stats
import matplotlib.pyplot as plt
from sklearn.metrics import f1_score, accuracy_score
import time

def scores_for_ks(test_labels, knn_labels, ks):
    f1_weight = []
    f1_macro = []
    f1_micro = []
    acc = []
    for k in ks:
        pred_k = stats.mode(knn_labels[:,:k], axis=1)[0].reshape((-1,))
        f1_weight.append(f1_score(test_labels, pred_k, average='weighted'))
        f1_macro.append(f1_score(test_labels, pred_k, average='macro'))
        f1_micro.append(f1_score(test_labels, pred_k, average='micro'))
        acc.append(accuracy_score(test_labels, pred_k))
    return {'f1_weight': f1_weight, 'f1_macro': f1_macro, 'f1_micro': f1_micro, 'accuracy': acc, }

def plot_prediction_quality(scores, ks):
    colors = ['r-', 'b-', 'g-','y-'][:len(scores)]
    for (k,v), color in zip(scores.items(), colors):
        plt.plot(ks, v, color, label=k)
    plt.legend()
    plt.xlabel('k')
    plt.ylabel('prediction quality')
    plt.show()

def evaluate_quality(predictor, test_features, test_labels, model_name, verbose=True, num_batches=100):
    """
    Evaluate quality metrics of a model on a test set. 
    """
    # tune the predictor to provide the verbose response
    predictor.accept = 'application/json; verbose=true'
    
    # split the test data set into num_batches batches and evaluate using prediction endpoint. 
    print('running prediction (quality)...')
    batches = np.array_split(test_features, num_batches)
    knn_labels = []
    for batch in batches:
        pred_result = predictor.predict(batch)
        cur_knn_labels = np.array([pred_result['predictions'][i]['labels'] for i in range(len(pred_result['predictions']))])
        knn_labels.append(cur_knn_labels)
    knn_labels = np.concatenate(knn_labels)
    print('running prediction (quality)... done')

    # figure out different k values
    top_k = knn_labels.shape[1]
    ks = range(1, top_k+1)
    
    # compute scores for the quality of the model for each value of k
    print('computing scores for all values of k... ')
    quality_scores = scores_for_ks(test_labels, knn_labels, ks)
    print('computing scores for all values of k... done')
    if verbose:
        plot_prediction_quality(quality_scores, ks)
    
    return quality_scores

def evaluate_latency(predictor, test_features, test_labels, model_name, verbose=True, num_batches=100):
    """
    Evaluate the run-time of a model on a test set.
    """
    # tune the predictor to provide the non-verbose response
    predictor.accept = 'application/json'
    
    # latency for large batches:
    # split the test data set into num_batches batches and evaluate the latencies of the calls to endpoint. 
    print('running prediction (latency)...')
    batches = np.array_split(test_features, num_batches)
    test_preds = []
    latency_sum = 0
    for batch in batches:
        start = time.time()
        pred_batch = predictor.predict(batch)
        latency_sum += time.time() - start
    latency_mean = latency_sum / float(num_batches)
    avg_batch_size = test_features.shape[0] / num_batches
    
    # estimate the latency for a batch of size 1
    latencies = []
    attempts = 2000
    for i in range(attempts):
        start = time.time()
        pred_batch = predictor.predict(test_features[i].reshape((1,-1)))
        latencies.append(time.time() - start)

    latencies = sorted(latencies)
    latency1_mean = sum(latencies) / float(attempts)
    latency1_p90 = latencies[int(attempts*0.9)]
    latency1_p99 = latencies[int(attempts*0.99)]
    print('running prediction (latency)... done')
    
    if verbose:
        print("{:<11} {:.3f}".format('Latency (ms, batch size %d):' % avg_batch_size, latency_mean * 1000))
        print("{:<11} {:.3f}".format('Latency (ms) mean for single item:', latency1_mean * 1000))
        print("{:<11} {:.3f}".format('Latency (ms) p90 for single item:', latency1_p90 * 1000))
        print("{:<11} {:.3f}".format('Latency (ms) p99 for single item:', latency1_p99 * 1000))
        
    return {'Latency': latency_mean, 'Latency1_mean': latency1_mean, 'Latency1_p90': latency1_p90, 
            'Latency1_p99': latency1_p99}

def evaluate(predictor, test_features, test_labels, model_name, verbose=True, num_batches=100):
    eval_result_q = evaluate_quality(pred, test_features, test_labels, model_name=model_name, verbose=verbose, num_batches=num_batches)
    eval_result_l = evaluate_latency(pred, test_features, test_labels, model_name=model_name, verbose=verbose, num_batches=num_batches)
    return dict(list(eval_result_q.items()) + list(eval_result_l.items()))
    
    

We are now ready to set up the endpoints. The following code will set up 4 endpoints, one for every option of cpu/gpu machine and Flat/IVFPQ index. For a gpu machine we use *ml.p2.xlarge* and for cpu we use *ml.c5.xlarge*. For cases where latency is not an issue, or the dataset is very small, we recommend using *ml.m4.xlarge* machines, as they are cheaper than the machines mentioned above. For the purpose of these notebook, we restrict our attention to compute optimized machines as we are also optimizing for latency. After setting an endpoint we evaluate it and then delete it. If you requires to keep using an endpoint, do not delete it until you are done with it.

In [ ]:
import time

instance_types = ['ml.c5.xlarge', 'ml.p2.xlarge']
index2estimator = {'flat_l2': knn_estimator_flat_l2, 'ivfpq_l2': knn_estimator_ivfpq_l2,
                  'flat_l2_large': knn_estimator_flat_l2_large, 'ivfpq_l2_large': knn_estimator_ivfpq_l2_large}

eval_results = {}

for index in index2estimator:
    estimator = index2estimator[index]
    eval_results[index] = {}
    for instance_type in instance_types:
        model_name = 'knn_%s_%s'%(index, instance_type)
        endpoint_name = 'knn-latency-%s-%s-%s'%(index.replace('_','-'), instance_type.replace('.','-'),
                                               str(time.time()).replace('.','-'))
        print('\nsetting up endpoint for instance_type=%s, index_type=%s' %(instance_type, index))
        pred = predictor_from_hyperparams(estimator, index, instance_type, endpoint_name=endpoint_name)
        print('')
        eval_result = evaluate(pred, test_features, test_labels, model_name=model_name, verbose=True)        
        eval_result['instance'] = instance_type 
        eval_result['index'] = index 
        eval_results[index][instance_type] = eval_result
        delete_endpoint(pred)

Looking at the plots it seems that the performance is best for small values of k. Let's view some values in a table to get a clear comparison. Below is a table with rows representing the models and columns, the different measurements. We highlight for every model any accuracy score that is at most 0.25% (relative) away from the maximum accuracy among the different values of k

In [ ]:
import pandas as pd

k_range = range(1, 13)
df_index = []
data = []
columns_lat = ['latency1K', 'latency1_mean', 'latency1_p90', 'latency1_p99']
columns_acc = ['acc_%d' % k for k in k_range]
columns = columns_lat + columns_acc

for index, index_res in eval_results.items():
    for instance, res in index_res.items():
        # for sample size?
        df_index.append(index+'_'+instance)
        latencies = np.array([res['Latency'], res['Latency1_mean'], res['Latency1_p90'], res['Latency1_p99']])
        row = np.concatenate([latencies*10,
                             res['accuracy'][k_range[0] - 1:k_range[-1] ]])
        row *= 100
        data.append(row)

df = pd.DataFrame(index=df_index, data=data, columns=columns)
df_acc = df[columns_acc]
df_lat = df[columns_lat]

def highlight_apx_max(row):
    '''
    highlight the aproximate best (max or min) in a Series yellow.
    '''
    max_val = row.max()
    colors = ['background-color: yellow' if cur_val >= max_val * 0.9975 else '' for cur_val in row]
        
    return colors

df_acc.round(decimals=1).style.apply(highlight_apx_max, axis=1)

Let's review the latencies. We'll highlight the latencies that are much worse (20% more) than the median value

In [ ]:
def highlight_far_from_min(row):
    '''
    highlight the aproximate best (max or min) in a Series yellow.
    '''
    med_val = row.median()
    colors = ['background-color: red' if cur_val >= med_val * 1.2 else '' for cur_val in row]
        
    return colors

df_lat.round(decimals=1).style.apply(highlight_far_from_min, axis=0)

#### Accuracy

For optimizing accuracy, the results show that for some of the models, k=1 yields the best results. We may want to play it safe and choose a larger value for k. The reason is that if the data slightly changes over time, the results associated with small values of k tend to change faster than those of larger values of k. A reasonable compromise could be choosing k=5. Not surprisingly, the accuracy becomes better if we choose a larger training set and if we use a brute-force index. This leads to 97% accuracy. For comparison, a linear model trained on the covtype dataset achieves roughly 72% accuracy. This is a pretty solid demonstration of the power of the kNN classifier.
If we are willing to get slightly less accurate results but at higher speed, we could choose either a large dataset with an approximate index **IVFPQ_l2_large_XX**, or a smaller sample size **flat_l2_ml_XX**. Both achieve roughly 93%-95% precision, yet have much more favorable latency scores.


#### Latency

We can see a single-query mean latency of under 10ms in most setting and under 20ms in the remaining. Some of this can likely be attributed to typical system overheads, so results may become better over time. One exception is for **flat_l2_large_ml.c5.xlarge**, where we pay for using an exact solution on a large dataset by having a latency of about 25ms. There is much more variance in the latency for a batch of roughly 1K points as the relative overhead there is lower. There, the slowest yet most accurate model requires roughly 360ms to return an answer while for the approximate versions we get over X5 speedup. If one would like to have the best accuracy but reduce latency, another option is to use a GPU machine. **flat_l2_large_ml.p2.xlarge** enjoys the best accuracy while keeping the latency for 1K points at less than 90ms and singe-query p90 latency at 14.2ms. The downside is the dollar cost, since a *ml.p2.xlarge* machine is more expensive than a *ml.c5.xlarge* machine. 


## Concluding Remarks

We've seen how to both train and host an inference endpoint for kNN. We've shown the ease of tuning a kNN algorithm and how to experiment with the different parameters, both those required at training and at inference. The experiment on the covtype dataset demonstrates the power of the simple procedure of kNN, especially when considering the final accuracy score of 96.8%, say, compared to linear model that achieves 72.8% accuracy. We explored the tradeoffs between an approximate index, subsampling the data, and using high vs low cost machines. The answer is case dependent and should fit the needs of the particular setup. 